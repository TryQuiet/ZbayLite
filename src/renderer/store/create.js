import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const composer = (enhancers) => {
  if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      process.env.NODE_ENV === 'development') {
    return window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__(enhancers)
  }
  return compose(enhancers)
}

export default ({
  initialState
} = {}) => createStore(
  reducers,
  initialState,
  composer(applyMiddleware(...[errorsMiddleware, createDebounce(), thunk, promise()]))
)
