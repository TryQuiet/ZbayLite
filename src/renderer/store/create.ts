import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'
import createSagaMiddleware from 'redux-saga'
import rootSaga from '../sagas/index.saga'
import { composeWithDevTools } from 'redux-devtools-extension'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const sagaMiddleware = createSagaMiddleware()

export default (initialState = {}) => {
  const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(
      applyMiddleware(...[errorsMiddleware, createDebounce(), sagaMiddleware, thunk, promise()])
    )
  )
  sagaMiddleware.run(rootSaga)
  return store
}
