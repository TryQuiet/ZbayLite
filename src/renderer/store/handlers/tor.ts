import { produce, immerable } from 'immer'
import net from 'net'
import { ipcRenderer } from 'electron'

import { createAction, handleActions } from 'redux-actions'
import torSelectors from '../selectors/tor'
import { successNotification } from './utils'
import notificationsHandlers from './notifications'
import electronStore from '../../../shared/electronStore'
import { actionTypes } from '../../../shared/static'

import { ActionsType, PayloadType } from './types'

export const client = new net.Socket()
export const defaultTorUrlProxy = 'localhost:9050'

class Tor {
  url: string
  enabled: boolean
  error?: string
  status: string

  constructor(values?: Partial<Tor>) {
    Object.assign(this, values)
    this[immerable] = true
  }
}

export const initialState: Tor = {
  ...new Tor({
    url: '',
    enabled: false,
    error: null,
    status: ''
  })
}

const setEnabled = createAction<{ enabled: boolean }>(actionTypes.SET_TOR_ENABLED)
const setUrl = createAction<{ url: string }>(actionTypes.SET_TOR_URL)
const setError = createAction<{ error: string }>(actionTypes.SET_TOR_ERROR)
const setStatus = createAction<{ status: string }>(actionTypes.SET_TOR_STATUS)

export const actions = {
  setUrl,
  setEnabled,
  setError,
  setStatus
}

export type TorActions = ActionsType<typeof actions>

let init = false
const initEvents = () => async dispatch => {
  client.on('error', () => {
    client.destroy()
    dispatch(setStatus({ status: 'down' }))
    dispatch(setError({ error: 'Cannot establish a connection' }))
  })
  client.on('data', data => {
    client.destroy()
    if (Buffer.from(data).toString('hex') === '0500') {
      dispatch(setStatus({ status: 'stable' }))
      dispatch(setError({ error: '' }))
    }
  })
}
const checkDeafult = () => async dispatch => {
  if (!init) {
    init = true
    await dispatch(initEvents())
  }
  const url = defaultTorUrlProxy.split(':')

  const checkedUrl = defaultTorUrlProxy
  dispatch(setUrl({ url: checkedUrl }))
  client.connect(Number(url[1]), url[0], () => {
    const msg = Buffer.from('050100', 'hex')
    client.write(msg)
  })
}

const checkTor = () => async (dispatch, getState) => {
  const tor = torSelectors.tor(getState())
  const url = tor.url.split(':')
  dispatch(setStatus({ status: 'loading' }))
  setTimeout(() => {
    if (torSelectors.tor(getState()).status === 'loading') {
      dispatch(setStatus({ status: 'down' }))
      dispatch(setError({ error: 'Timeout error' }))
    }
  }, 5000)

  client.connect(Number(url[1]), url[0], () => {
    const msg = Buffer.from('050100', 'hex')
    client.write(msg)
  })
}
export const createZcashNode = torUrl => async dispatch => {
  electronStore.set('torEnabled', !!torUrl)
  let ipAddress
  if (torUrl?.startsWith('localhost')) {
    ipAddress = torUrl.replace('localhost', '127.0.0.1')
  } else {
    ipAddress = torUrl
  }
  ipcRenderer.send('create-node', ipAddress)
  if (torUrl) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({ message: 'You are using Tor proxy.' })
      )
    )
  }
}
export const epics = {
  checkTor,
  createZcashNode,
  checkDeafult
}

export const reducer = handleActions<Tor, PayloadType<TorActions>>(
  {
    [setEnabled.toString()]: (state, { payload: { enabled } }: TorActions['setEnabled']) =>
      produce(state, draft => {
        if (enabled) {
          draft.enabled = enabled
          draft.status = 'down'
        } else {
          draft.enabled = enabled
          draft.status = 'down'
          draft.error = ''
          draft.url = ''
        }
      }),
    [setUrl.toString()]: (state, { payload: { url } }: TorActions['setUrl']) =>
      produce(state, draft => {
        draft.url = url
        draft.status = 'down'
      }),
    [setError.toString()]: (state, { payload: { error } }: TorActions['setError']) =>
      produce(state, draft => {
        draft.error = error
      }),
    [setStatus.toString()]: (state, { payload: { status } }: TorActions['setStatus']) =>
      produce(state, draft => {
        draft.status = status
      })
  },
  initialState
)
export default {
  actions,
  epics,
  reducer
}
