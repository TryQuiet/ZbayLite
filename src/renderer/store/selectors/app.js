import { createSelector } from 'reselect'

const store = s => s

export const app = createSelector(store, state => state.get('app'))

const version = createSelector(app, a => a.version)
const useTor = createSelector(app, a => a.useTor)
const transfers = createSelector(app, a => a.transfers)
const currentModalTab = createSelector(app, a => a.modalTabToOpen)
const allTransfersCount = createSelector(app, a => a.allTransfersCount)
const newTransfersCounter = createSelector(app, a => a.newTransfersCounter)
const isInitialLoadFinished = createSelector(app, a => a.isInitialLoadFinished)
const directMessageQueueLock = createSelector(
  app,
  a => a.directMessageQueueLock
)
const messageQueueLock = createSelector(app, a => a.messageQueueLock)

export default {
  version,
  transfers,
  currentModalTab,
  allTransfersCount,
  newTransfersCounter,
  messageQueueLock,
  directMessageQueueLock,
  isInitialLoadFinished,
  useTor
}
