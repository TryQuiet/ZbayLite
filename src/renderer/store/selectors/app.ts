import { createSelector } from 'reselect'

import { Store } from '../reducers'

const app = (s: Store) => s.app

const version = createSelector(app, (a) => a.version)
const currentModalTab = createSelector(app, (a) => a.modalTabToOpen)
const isInitialLoadFinished = createSelector(
  app,
  (a) => a.isInitialLoadFinished
)

export default {
  version,
  currentModalTab,
  isInitialLoadFinished
}
