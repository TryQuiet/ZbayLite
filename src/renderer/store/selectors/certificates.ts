import { Store } from '../reducers'
import { createSelector } from 'reselect'

const certificates = (s: Store) => s.certificates
const ownCertificate = createSelector(certificates, d => d.ownCertificate.certificate)

export default {
  certificates,
  ownCertificate
}
