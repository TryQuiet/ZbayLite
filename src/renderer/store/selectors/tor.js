import { createSelector } from 'reselect'
const store = s => s

const tor = createSelector(
  store,
  state => state.tor
)
const torEnabled = createSelector(
  tor,
  tor => tor.enabled
)

export default {
  tor,
  torEnabled
}
