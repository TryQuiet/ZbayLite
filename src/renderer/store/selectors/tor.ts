import { createSelector } from 'reselect'

import { Store } from '../reducers'

const tor = (s: Store) => s.tor

const torEnabled = createSelector(
  tor,
  tor => tor.enabled
)

export default {
  tor,
  torEnabled
}
