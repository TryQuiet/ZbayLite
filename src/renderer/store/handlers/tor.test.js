import create from '../create'
import { actions, initialState, epics, client } from './tor'
import selectors from '../selectors/tor'

jest.mock('../../../shared/electronStore', () => ({
  set: () => {},
  get: () => {}
}))

describe('Tor reducer handles ', () => {
  let store = null

  beforeEach(() => {
    store = create({
      initialState: {
        tor: { ...initialState }
      }
    })
    jest.clearAllMocks()
  })

  describe('actions', () => {
    it('- setEnabled', () => {
      store.dispatch(actions.setEnabled({ enabled: true }))
      expect(selectors.tor(store.getState())).toMatchSnapshot()
    })
    it('- setUrl', () => {
      store.dispatch(actions.setUrl({ url: 'testhost:9999' }))
      expect(selectors.tor(store.getState())).toMatchSnapshot()
    })
    it('- setError', () => {
      store.dispatch(actions.setError({ error: 'error' }))
      expect(selectors.tor(store.getState())).toMatchSnapshot()
    })
    it('- setStatus', () => {
      store.dispatch(actions.setStatus({ status: 'up' }))
      expect(selectors.tor(store.getState())).toMatchSnapshot()
    })
  })
  describe('epics', () => {
    it('- checkTor', async () => {
      const connectMock = jest.spyOn(client, 'connect').mockImplementation(() => {})

      await store.dispatch(epics.checkTor())
      expect(connectMock).toHaveBeenCalled()
    })
  })
})
