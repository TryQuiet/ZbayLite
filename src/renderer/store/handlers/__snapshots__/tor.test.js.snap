// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tor reducer handles  actions - setEnabled 1`] = `
Object {
  "enabled": true,
  "error": null,
  "status": "down",
  "url": "",
}
`;

exports[`Tor reducer handles  actions - setError 1`] = `
Object {
  "enabled": false,
  "error": "error",
  "status": "",
  "url": "",
}
`;

exports[`Tor reducer handles  actions - setStatus 1`] = `
Object {
  "enabled": false,
  "error": null,
  "status": "up",
  "url": "",
}
`;

exports[`Tor reducer handles  actions - setUrl 1`] = `
Object {
  "enabled": false,
  "error": null,
  "status": "down",
  "url": "testhost:9999",
}
`;
